### PR TITLE
Foreign Shells

### PR DESCRIPTION
**Pull request by scopatz**
This PR adds a generalized capability for pulling in information (environment and aliases) from foreign shells, such as Bash.  As a side effect, this also allows you to turn off loading in any data from foreign shells at all.  

This depends on #410 and #411 and was built to mitigate #403.

If @asmeurer and/or @auscompgeek could try this out on their machines and let me know whether it works for them, that'd be great!  I'll merge this in, if I get the OK from either of these people.

Also, note that this disables using login shells by default, though it can easily be reinstated.

### CODE REVIEW

**Comment by scopatz** [comment]
Here are some timings on one of my better machines:

```console
# master branch
scopatz@artemis ~ $ time -f "%E real\t%U user\t%S sys" xonsh -c exit
0:00.26 real	0.23 user	0.00 sys

# foreignsh branch, loading bash
scopatz@artemis ~ $ time -f "%E real\t%U user\t%S sys" xonsh -c exit
0:00.18 real	0.16 user	0.00 sys

# foreignsh branch, loading nothing
scopatz@artemis ~ $ time -f "%E real\t%U user\t%S sys" xonsh -c exit
0:00.15 real	0.14 user	0.00 sys

# bash
scopatz@artemis ~ $ time -f "%E real\t%U user\t%S sys" bash -c exit
0:00.00 real	0.00 user	0.00 sys

# python
scopatz@artemis ~ $ time -f "%E real\t%U user\t%S sys" python -c "exit()"
0:00.05 real	0.03 user	0.01 sys
```

File: xonsh/environ.py [hunk 1]
```diff
@@ -73,7 +168,7 @@ class Env(MutableMapping):
       locale via locale.getlocale() and locale.setlocale() functions.
 
     An Env instance may be converted to an untyped version suitable for
-    use in a subprocess.
+    use in a subprocess.
```
Comment (auscompgeek) [inline xonsh/environ.py]: Someone needs to set up their editor to strip trailing whitespace :wink:

File: xonsh/shell.py [hunk 2]
```diff
@@ -26,25 +26,26 @@ class Shell(object):
     def __init__(self, ctx=None, shell_type=None, **kwargs):
         self._init_environ(ctx)
         env = builtins.__xonsh_env__
-
+        # piclk a valid shell
```
Comment (auscompgeek) [inline xonsh/shell.py]: typo

File: xonsh/foreign_shells.py [hunk 3]
```diff
@@ -0,0 +1,243 @@
...
+        currenv = builtins.__xonsh_env__.detype()
+    elif currenv is not None:
+        currenv = dict(currenv)
+    try:
+        s = subprocess.check_output(cmd,stderr=subprocess.PIPE, env=currenv,
+                                    universal_newlines=True)
+    except (subprocess.CalledProcessError, FileNotFoundError):
+        if not safe:
+            raise
+        s = FAILED_COMMAND_STDOUT
```
Comment (auscompgeek) [inline xonsh/foreign_shells.py]: Could we just `return {}, {}` here instead?

File: docs/api/index.rst [hunk 4]
```diff
@@ -51,5 +51,6 @@ For those of you who want the gritty details.
     lazyjson
     teepty
     openpy
+    foriegn_shells
```
Comment (auscompgeek) [inline docs/api/index.rst]: typo

File: xonsh/foreign_shells.py [hunk 5]
```diff
@@ -0,0 +1,243 @@
...
+        Path to the static config file. Not compatible with shell not being None.
+        If both shell and config is None, then it will be read from the 
+        $XONSHCONFIG environment variable.
+    issue_warning : bool, optional
+        Issues warnings if config file cannot be found.
+
+    Returns
+    -------
+    env : dict
+        A dictionary of the merged environments.
```
Comment (auscompgeek) [inline xonsh/foreign_shells.py]: copy-paste fail?

**Comment by auscompgeek** [comment]
Will `source_bash()` be rewritten to use this?

File: xonsh/environ.py (hunk 1 above)
Comment (scopatz) [inline xonsh/environ.py]: Yeah, I haven't added that feature to xo yet, soon probably :)

File: xonsh/foreign_shells.py (hunk 3 above)
Comment (scopatz) [inline xonsh/foreign_shells.py]: yep.

**Comment by scopatz** [comment]
Thanks for the review @auscompgeek! I have addressed everything that you brought up in-line.

> Will source_bash() be rewritten to use this?

It probably should be, but I don't want this PR to get too big. Maybe you could open an issue for that?

**Comment by scopatz** [comment]
Also, does this actually work for you?

File: xonsh/foreign_shells.py [hunk 6]
```diff
@@ -0,0 +1,236 @@
...
+        Path to the static config file. Not compatible with shell not being None.
+        If both shell and config is None, then it will be read from the 
+        $XONSHCONFIG environment variable.
+    issue_warning : bool, optional
+        Issues warnings if config file cannot be found.
+
+    Returns
+    -------
+    aliases : dict
+        A dictionary of the merged alaiases.
```
Comment (auscompgeek) [inline xonsh/foreign_shells.py]: Maybe you should invest in an editor that does spellcheck...

**Comment by auscompgeek** [comment]
Seems to work well for me on my Arch Linux install. Although, do we have to run both `env` and `alias` when we only want one of the two?

File: xonsh/foreign_shells.py (hunk 6 above)
Comment (scopatz) [inline xonsh/foreign_shells.py]: Again, a feature I have considered adding to xo. PRs welcome.

**Comment by scopatz** [comment]
> Seems to work well for me on my Arch Linux install. Although, do we have to run both env and alias when we only want one of the two?

I can't think of a case where only one is desired. Maybe I am missing something though...  I'd almost certainly accept a PR that added getting one or the other as an optional feature. (Say you just want env, you could set aliascmd to false.) However, getting both simultaneously is fairly advantageous. 

As a historical note, originally env and aliases were obtained together.  Then someone put in a PR to split getting them up, but it nearly doubled the startup time for xonsh. Enough people have complained that startup was too slow to motivate this PR.

**Pull request merged**